python_modules/dagster/dagster/_core/execution/run_metrics_thread.py

```diff
@@ -0,0 +1,252 @@
+        }
+        stats_dict.update(gen_metrics)
+
+    return {**stats_dict, "python.runtime.gc_freeze_count": gc.get_freeze_count()}
+
+
+def _report_run_metrics(
+    instance: DagsterInstance,
+    dagster_run: DagsterRun,
+    metrics: Dict[str, float],
+    run_tags: Dict[str, str],
+):
+    datapoints: List[TelemetryDataPoint] = []
+    for metric, value in metrics.items():
+        if value is None:
+            continue
+        try:
+            datapoint = TelemetryDataPoint(
+                name=metric,
+                value=float(value),
+            )
+            datapoints.append(datapoint)
+        except ValueError:
+            logging.warning(f"Failed to convert metric value to float: {metric}={value}, skipping")
+
+    telemetry_data = RunTelemetryData(run_id=dagster_run.run_id, datapoints=datapoints)
+
+    instance._run_storage.add_run_telemetry(  # noqa: SLF001
+        telemetry_data, tags=run_tags
+    )
+
+
+def _capture_metrics(
+    instance: DagsterInstance,
+    dagster_run: DagsterRun,
+    container_metrics_enabled: bool,
+    python_metrics_enabled: bool,
+    shutdown_event: threading.Event,
+    polling_interval: float = DEFAULT_RUN_METRICS_POLL_INTERVAL_SECONDS,
+    logger: Optional[logging.Logger] = None,
+) -> bool:
+    check.inst_param(instance, "instance", DagsterInstance)
+    check.inst_param(dagster_run, "dagster_run", DagsterRun)
+    check.bool_param(container_metrics_enabled, "container_metrics_enabled")
+    check.bool_param(python_metrics_enabled, "python_metrics_enabled")
+    check.inst_param(shutdown_event, "shutdown_event", threading.Event)
+    check.float_param(polling_interval, "polling_interval")
+    check.opt_inst_param(logger, "logger", logging.Logger)
+
+    if not (container_metrics_enabled or python_metrics_enabled):
+        raise ValueError("No metrics enabled")
+
+    run_tags = _metric_tags(instance, dagster_run)
+
+    if logger:
+        logging.debug(f"Starting metrics capture thread with tags: {run_tags}")
+        logging.debug(f"  [container_metrics_enabled={container_metrics_enabled}]")
+        logging.debug(f"  [python_metrics_enabled={python_metrics_enabled}]")
+
+    while not shutdown_event.is_set():
+        try:
+            metrics = {}
+
+            if container_metrics_enabled:
+                container_metrics = _get_container_metrics(logger=logger)
+                metrics.update(container_metrics)
+
+            if python_metrics_enabled:
+                python_metrics = _get_python_runtime_metrics()
+                metrics.update(python_metrics)
+
+            if len(metrics) > 0:
+                _report_run_metrics(
+                    instance,
+                    dagster_run=dagster_run,
+                    metrics=metrics,
+                    run_tags=run_tags,
+                )
+
+        except:
+            logging.error(
+                "Exception during capture of metrics, will cease capturing", exc_info=True
+            )
```

Comment:
logging.exception("Exception during capture of metrics, will cease capturing")? (will automatically include the stack trace)